sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java

```diff
@@ -173,6 +173,12 @@ public DestinationState(
         this.useDefaultStream = useDefaultStream;
       }
 
+      void teardown() {
+        if (streamAppendClient != null) {
+          runAsyncIgnoreFailure(closeWriterExecutor, streamAppendClient::unpin);
```

Comment:
do we need to invalidate it from the cache as well here?

Reply:
No - unpin will only close the connection if close() has already been called. Since the client is still  healthy, we can leave  it in the cache